keps/sig-scheduling/5278-nominated-node-name-for-expectation/README.md

```diff
@@ -239,68 +240,117 @@ If we can keep where the pod was going to go at `NominatedNodeName`, the schedul
-Kubernetes ecosystem.
+#### NominatedNodeName can already be set by other components now.
 
-How will security be reviewed, and by whom?
+There aren't any guardrails preventing other components from setting NominatedNodeName now.
+In such cases, the semantic is not well defined now and the outcome of it may not match user
+expectations.
 
-How will UX be reviewed, and by whom?
+This KEP is a step towards clarifying this semantic instead of maintaining status-quo.
+
+#### Confusing semantics of `NominatedNodeName`
+
+Up until now, `NominatedNodeName` was expressing the decision made by scheduler to put a given
+pod on a given node, while waiting for the preemption. The decision could be changed later so
+it didn't have to be a final decision, but it was describing the "current plan of record".
+
+If we put more components into the picture (e.g. ClusterAutoscaler and Karpenter), we effectively
+get a more complex state machine, with the following states:
+
+1. pending pod
+1. pod proposed to node (by external component) [not approved by scheduler]
+1. pod nominted to node (based on external proposal) and waiting for node (e.g. being created & ready)
+1. pod nominated to node and waiting for preemption
+1. pod allocated to node and waiting for binding
+1. pod bound
+
+The important part is that if we decide to use `NominatedNodeName` to store all that information,
+we're effectively losing the ability to distinguish between those states.
+
+We may argue that as long as the decision was made by the scheduler, the exact reason and state
+probably isn't that important - the content of `NominatedNodeName` can be interpreted as
+"current plan of record for this pod from scheduler perspective".
+
+But the `pod proposed to node` state is visibly different. In particular external components
+may overallocate the pods on the node, those pods may not match scheduling constraints etc.
+We can't claim that it's a current plan of record of the scheduler. It's a hint that we want
+scheduler to take into account.
+
+As a result, we would like to distinguish that state. The potential options include:
+- introducing a new field (e.g. `NodeNameHint` or sth like that for that purpose)
+- using `pod.ObjectMeta.ManagedFields` to detect who set the field
+
+Given that, even as of now arbitrary component or actor can already set `NNN`, we suggest to
+introduce a given rule of thumb:
+- if `NominatedNodeName` was set by scheduler (based on ManagedFields), no external component
+should try to modify it
+- if `NominatedNodeName` was set by an external component (based on ManagedFields), it should
+be treated as a hint to scheduler and not as an official scheduler's plan of record
+- if `NominatedNodeName` was set by an external component (based on ManagedFields), this
+component should clear or update it (using PUT or APPLY to ensure it won't conflict with
+potentil update from scheduler) to reflect the new hint
```

Comment:
@dom4ha - in the context of different phases of reservation/allocation of resources that we were talking about, I think that what I'm proposing above paths they way towards additive extension in the future

With such approach, we're describing how the semantics should work, and eventually we can add `stage` field that would reflect that more explicitly.